pennylane/beta/tapes/tape.py

```diff
@@ -936,3 +936,387 @@ def execute_device(self, params, device):
+        }
+
+        # check and raise an error if any parameters are non-differentiable
+        nondiff_params = {idx for idx, g in allowed_param_methods.items() if g is None}
+
+        if nondiff_params:
+            raise ValueError(f"Cannot differentiate with respect to parameter(s) {nondiff_params}")
+
+        numeric_params = {idx for idx, g in allowed_param_methods.items() if g == "F"}
+
+        if method == "analytic":
+            # If explicitly using analytic mode, ensure that all parameters
+            # support analytic differentiation.
+
+            if numeric_params:
+                raise ValueError(
+                    f"The analytic gradient method cannot be used with the argument(s) {numeric_params}."
+                )
+
+        return tuple(allowed_param_methods.values())
+
+    def numeric_pd(self, idx, device, params=None, **options):
+        """Evaluate the gradient of the tape with respect to
+        a single trainable tape parameter using numerical finite-differences.
+
+        Args:
+            idx (int): trainable parameter index to differentiate with respect to
+            device (~.Device, ~.QubitDevice): a PennyLane device
+                that can execute quantum operations and return measurement statistics
+            params (list[Any]): The quantum tape operation parameters. If not provided,
+                the current tape parameter values are used (via :meth:`~.get_parameters`).
+
+        Keyword Args:
+            h=1e-7 (float): finite difference method step size
+            order=1 (int): The order of the finite difference method to use. ``1`` corresponds
+                to forward finite differences, ``2`` to centered finite differences.
+
+        Returns:
+            array[float]: 1-dimensional array of length determined by the tape output
+                measurement statistics
+        """
+        if params is None:
+            params = np.array(self.get_parameters())
+
+        order = options.get("order", 1)
+        h = options.get("h", 1e-7)
+
+        shift = np.zeros_like(params)
+        shift[idx] = h
+
+        if order == 1:
+            # Forward finite-difference.
+            # Check if the device has already be pre-computed with
+            # unshifted parameter values, to avoid redundant evaluations.
+            y0 = options.get("y0", None)
+
+            if y0 is None:
+                y0 = np.asarray(self.execute_device(params, device))
+
+            y = np.array(self.execute_device(params + shift, device))
+            return (y - y0) / h
+
+        if order == 2:
+            # central finite difference
+            shift_forward = np.array(self.execute_device(params + shift / 2, device))
+            shift_backward = np.array(self.execute_device(params - shift / 2, device))
+            return (shift_forward - shift_backward) / h
+
+        raise ValueError("Order must be 1 or 2.")
+
+    def device_pd(self, device, params=None, **options):
```

Comment:
I guess this is forward thinking, in case at some point we _do_ want to pass options to the device here.

Easier to include this now (even with no options), than to have to refactor the devices later to add it